[RFC] vim-patch:8.0.{1020,1048},8.1.0052

janlazo opened the pull request:
**vim-patch:8.0.1020: when a timer calls getchar(1) input is overwritten**

Problem:    When a timer calls getchar(1) input is overwritten.
Solution:   Increment tb_change_cnt in inchar(). (closes vim/vim#1940)
vim/vim@0f0f230

**vim-patch:8.0.1048: no test for what 8.0.1020 fixes**

Problem:    No test for what 8.0.1020 fixes.
Solution:   Add test_feedinput().  Add a test. (Ozaki Kiichi, closes vim/vim#2046)
vim/vim@5e80de3

**vim-patch:8.1.0052: when mapping to <Nop> times out the next mapping is skipped**

Problem:    When a mapping to <Nop> times out the next mapping is skipped.
Solution:   Reset "timedout" when waiting for a character. (Christian
            Brabandt, closes vim/vim#2921)
vim/vim@83f4cbd

I omit code for `test_feedinput` so 8.0.1048 includes the test code only.

janlazo committed:
vim-patch:8.0.1020: when a timer calls getchar(1) input is overwritten

Problem:    When a timer calls getchar(1) input is overwritten.
Solution:   Increment tb_change_cnt in inchar(). (closes vim/vim#1940)
https://github.com/vim/vim/commit/0f0f230012f5a9beb6876158a17b432534836c6f

janlazo committed:
lint

janlazo committed:
vim-patch:8.0.1048: no test for what 8.0.1020 fixes

Problem:    No test for what 8.0.1020 fixes.
Solution:   Add test_feedinput().  Add a test. (Ozaki Kiichi, closes vim/vim#2046)
https://github.com/vim/vim/commit/5e80de3f3e572805fe734b66bc42c13303ad9bdb

justinmk commented:
> test_feedinput so 8.0.1048 includes the test code only

We should probably add a screen test for that, in `test/functional/eval/timer_spec.lua` . Maybe this test just needs an extra step: https://github.com/neovim/neovim/blob/a466bd2927c9d15071aba425a01f11435ae4bbd2/test/functional/eval/timer_spec.lua#L88

janlazo commented:
Is that the fragile test failing in Quickbuild? It doesn't happen now but tests using timer + `getchar()` regularly failed in my previous PRs if I didn't touch code for timers.

janlazo committed:
vim-patch:8.1.0052: when mapping to <Nop> times out the next mapping is skipped

Problem:    When a mapping to <Nop> times out the next mapping is skipped.
Solution:   Reset "timedout" when waiting for a character. (Christian
            Brabandt, closes vim/vim#2921)
https://github.com/vim/vim/commit/83f4cbd973731872b633d6ba0caf850fb708d70c

janlazo committed:
tests: call getchar(1) in timer callback

Test changes from 8.0.1020 and 8.0.1048.

janlazo commented on test/functional/eval/timer_spec.lua:
```diff
@@ -96,6 +96,7 @@ describe('timers', function()
     source([[
       func! AddItem(timer)
         call nvim_buf_set_lines(0, 2, 2, v:true, ['ITEM 3'])
+        call getchar(1)
```
I added this to match `FeedAndPeek()` in `test_timers.vim`. I don't know what to do with the second timer that calls `Interrupt()`.
